### PR TITLE
feat(electron): shared storage

### PR DESCRIPTION
**Pull request by EYHN**

### CODE REVIEW

**Comment by EYHN** [comment]
* **#7440** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7440?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#7509](https://github.com/toeverything/AFFiNE/pull/7509) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7509?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#7635](https://github.com/toeverything/AFFiNE/pull/7635) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7635?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#7492** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7492?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#7491** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7491?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EYHN and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by pengx17 — APPROVED** [review]
LGTM

File: packages/frontend/electron/src/main/shared-storage/json-file.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import { debounceTime, Observable, timeout } from 'rxjs';
+
+export class PersistentJSONFileStorage implements Memento {
+  data: Record<string, any> = {};
+  subscriptions: Map<string, Set<(p: any) => void>> = new Map();
+  subscriptionAll: Set<(p: Record<string, any>) => void> = new Set();
+
+  constructor(readonly filepath: string) {
+    try {
+      this.data = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
```
Comment (pengx17) [inline packages/frontend/electron/src/main/shared-storage/json-file.ts:21]: better check if file exists before reading. no need to throw if it is not being created yet.

File: packages/frontend/electron/src/main/shared-storage/json-file.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    } catch (err) {
+      console.error('failed to load file', err);
+    }
```
Comment (pengx17) [inline packages/frontend/electron/src/main/shared-storage/json-file.ts:34]: use logger from '../loger'

File: packages/frontend/electron/src/main/shared-storage/json-file.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            'utf-8'
+          );
+        } catch (err) {
+          console.error(`failed to save file, ${this.filepath}`, err);
+        }
+      }).pipe(
+        timeout(5000),
+        backoffRetry({
+          count: Infinity,
+        })
```
Comment (pengx17) [inline packages/frontend/electron/src/main/shared-storage/json-file.ts]:
```suggestion
        ,
        mergeMap(() => EMPTY)
```

File: packages/frontend/electron/src/main/shared-storage/json-file.ts (hunk 1 above)
Comment (EYHN) [reply packages/frontend/electron/src/main/shared-storage/json-file.ts:21]: https://nodejs.org/docs/latest/api/fs.html#fsexistspath-callback

CleanShot 2024-07-15 at 10 26 40

It is best practice to read the file directly.

File: packages/frontend/electron/src/main/shared-storage/json-file.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+export class PersistentJSONFileStorage implements Memento {
+  data: Record<string, any> = {};
+  subscriptions: Map<string, Set<(p: any) => void>> = new Map();
+  subscriptionAll: Set<(p: Record<string, any>) => void> = new Set();
+
+  constructor(readonly filepath: string) {
+    try {
+      this.data = JSON.parse(fs.readFileSync(filepath, 'utf-8'));
+    } catch (err) {
+      console.error('failed to load file', err);
```
Comment (pengx17) [inline packages/frontend/electron/src/main/shared-storage/json-file.ts]: 👌
Please ignore file not exist error here

**Comment by EYHN** [comment]
### Merge activity

* **Jul 14, 10:49 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/7492) rebased this pull request after merging its parent, because this pull request is set to merge when ready.
* **Jul 14, 11:04 PM EDT**: [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/7492) rebased this pull request after merging its parent, because this pull request is set to merge when ready.
* **Jul 14, 11:20 PM EDT**: `EYHN` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE).
* **Jul 14, 11:36 PM EDT**: `EYHN`  merged this pull request with the [Graphite merge queue](https://app.graphite.dev/merges?org=toeverything&repo=AFFiNE).

**Pull request merged**